File: arrays/three_sum/go/solution_no_sort.go
package main

import (
	"sort"
	"strconv"
	"strings"
)

func threeSumNoSort(nums []int) [][]int {
	uniq := make(map[string]bool)
	dups := make(map[int]bool)
	seen := make(map[int]int)

	for i, firstNum := range nums {
		_, ok := dups[firstNum]
		if ok {
			continue
		}

		dups[firstNum] = true

		for j := i + 1; j < len(nums); j++ {
			secondNum := nums[j]

			// Высчитываем третье искомое число
			thirdNum := 0 - firstNum - secondNum

			idx, ok := seen[thirdNum]

			if ok && idx == i {
				// Собираем уникальную тройку
				triplet := []string{strconv.Itoa(firstNum), strconv.Itoa(secondNum), strconv.Itoa(thirdNum)}

				// Сортируем тройку, чтобы избежать дублей
				sort.Strings(triplet)

				// Конвертируем тройку в строку и добавляем в map
				str := strings.Join(triplet, ".")
				uniq[str] = true
			}

			seen[secondNum] = i
		}
	}

	// Разбираем map с уникальными ключами обратно в слайс троек
	res := make([][]int, 0, len(uniq))

	for key := range uniq {
		numStrs := strings.Split(key, ".")
		resNums := make([]int, 0, len(numStrs))

		for _, numStr := range numStrs {
			num, _ := strconv.Atoi(numStr)
			resNums = append(resNums, num)
		}
		res = append(res, resNums)
	}

	return res
}

File: _posts/arrays/three_sum/ts/solution_no_sort.ts
export const threeSumNoSort = (nums: number[]): number[][] => {
    const uniq = new Set<string>()
    const dups = new Set<number>()
    const seen = new Map<number, number>()

    for (let i = 0; i < nums.length; i++) {
        const firstNum = nums[i]

        if (dups.has(firstNum)) {
            continue
        }

        dups.add(firstNum)

        for (let j = i + 1; j < nums.length; j++) {
            const secondNum = nums[j]

            // Высчитываем третье искомое число
            const thirdNum = 0 - firstNum - secondNum

            if (seen.has(thirdNum) && seen.get(thirdNum) === firstNum) {
                // Собираем уникальную тройку
                const triplet = [firstNum, secondNum, thirdNum]

                // Сортируем тройку, чтобы избежать дублей
                triplet.sort()

                // Конвертируем тройку в строку и добавляем в set
                uniq.add(triplet.join('.'))
            }

            seen.set(secondNum, firstNum)
        }
    }

    // Конвертируем set строк в массив
    return Array.from(uniq).reduce<number[][]>((acc, tripletStr) => {
        const tripletArr = tripletStr.split('.')
        const triplet = tripletArr.map(numStr => Number(numStr))
        return [...acc, triplet]
    }, [])
}